visualizations.py

```diff
@@ -123,13 +134,124 @@ def display_embedding_graph(model_builder):
 
     # Add xticks on the middle of the group bars
-    plt.xlabel('Answer Number', fontweight='bold')
+    plt.xlabel('Embedding Dimension', fontweight='bold')
     plt.xticks([r + barWidth + 1 for r in range(len(bars1))],
                np.arange(1, len(bars1) + 1))
 
     plt.ylabel('Embedding Value', fontweight='bold')
     plt.title('Answer Embeddings', fontweight='bold')
 
-    # Create legend 
+    
+    plt.legend()
+
+    if len(argv) > 1:
+        plt.savefig(argv[1] + "/embedding.png")
+    else:
+        plt.savefig("embedding.png")
+    plt.show()
+    
+
+
+def display_perturbation_vs_epsilon(model_builder, ranker, answer_num, ref_num, epsilons):
+    '''
+    Calculates the difference between unperturbed input rank and fgsm/randomly 
+    perturbed input over a variety of epsilon values.
+
+    Args:
+        model_builder: The model in use.
+        ranker: Ranking object.
+        answer_num: The answer that is perturbed.
+        ref_num: The reference answer for the perturbed answer.
+        epsilons: Array containing the different values for epsilon.
+
+    Returns:
+        amount_of_perturbation_fgsm: Array containing amount of 
+        perturbation for FGSM perturbed input.
+        amount_of_perturbation_random: Array containing amount of 
+        perturbation for randomly perturbed input.
+
+    '''
+    amount_of_perturbation_fgsm = []
+    amount_of_perturbation_random = []
+    
+
+    init_variables(model_builder, ranker, constants._TEST_DATA_PATH,
+                   answer_num, epsilons[0], ref_num, True)
+    for perturb in epsilons:
+        init_variables(
+            model_builder,
+            ranker,
+            constants._TEST_DATA_PATH,
+            answer_num,
+            perturb,
+            ref_num,
+            False)
+        change_in_rank_fgsm = abs(
+            model_builder.ranking_array[answer_num] -
+            model_builder.perturbed_ranking_array[answer_num])
+        amount_of_perturbation_fgsm.append(change_in_rank_fgsm)
+
+        change_in_rank_random = abs(
+            model_builder.ranking_array[answer_num] -
+            model_builder.random_ranking_array[answer_num])
+        amount_of_perturbation_random.append(change_in_rank_random)
+    return amount_of_perturbation_fgsm, amount_of_perturbation_random
+
+
+def perturbation_vs_epsilon_graph(model_builder, ranker, answer_num, ref_num, argv):
+    """
+    Creates and saves bar graph showing difference in perturbation for both
+    fgsm-perturbed and random-perturbed input for a variety of epsilon values.
+
+    Args:
+        model_builder: The model in use.
+        ranker: Ranking object:
+        answer_num: The answer that is perturbed.
+        ref_num: The reference answer for the perturbed answer.
+        argv: Command line input to be used for directory name.
+    """
+    #range of epsilons can be modified as desired to change
+    #number of bars or increment size
+    epsilons = np.arange(.1, .5, .1)

```

Comment:
either declare this as a constant at the top of the file or make it configurable via a command line parameter. Avoid magic numbers such as this appearing unexplained in the code.